rasa/nlu/training_data/formats/markdown.py

```diff
@@ -211,25 +212,31 @@ def dumps(self, training_data: "TrainingData") -> Text:
     def _generate_training_examples_md(self, training_data: "TrainingData") -> Text:
         """Generates markdown training examples."""
 
-        # Sort training data by intent
-        training_examples = sorted(
-            [e.as_dict_nlu() for e in training_data.training_examples],
-            key=lambda k: k[MESSAGE_INTENT_ATTRIBUTE],
-        )
-        md = ""
-        for i, example in enumerate(training_examples):
-            intent = training_examples[i - 1][MESSAGE_INTENT_ATTRIBUTE]
-            if i == 0 or intent != example[MESSAGE_INTENT_ATTRIBUTE]:
-                md += self._generate_section_header_md(
-                    INTENT,
-                    example[MESSAGE_INTENT_ATTRIBUTE],
-                    example.get(MESSAGE_RESPONSE_KEY_ATTRIBUTE, None),
-                    i != 0,
-                )
+        training_examples = OrderedDict()
 
-            md += self._generate_item_md(self._generate_message_md(example))
+        # Sort by intent while keeping basic intent order
+        for example in [e.as_dict_nlu() for e in training_data.training_examples]:
+            intent = example[MESSAGE_INTENT_ATTRIBUTE]
+            training_examples.setdefault(intent, [])
+            training_examples[intent].append(example)
 
-        return md
+        # Don't prepend newline for first line
+        prepend_newline = False
+        lines = []
+
+        for intent, examples in training_examples.items():
+            section_header = self._generate_section_header_md(
```

Comment:
can we:
* collect `sections` instead of lines (one item per section)
* `"\n".join(sections)` in L239
* remove the `prepend_newline` arg to this function as well as the local variable?

Reply:
just took a look and this is used in a couple other functions as well, so would be a larger refactoring - happy to merge as is.

Reply:
yep, that's why I didn't do it 😄